CircularOptionPicker: refactor to TypeScript

chad1008 opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Refactor `CircularOptionPicker` component to TypeScript

Part of https://github.com/WordPress/gutenberg/issues/35744
Co-authored with @ciampo

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
The refactor to TypeScript has many benefits (auto-generated docs, static linting and error prevention, better IDE experience). See https://github.com/WordPress/gutenberg/issues/35744 for more details

## How?
Followed the steps in the [TypeScript migration guide](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md?rgh-link-date=2022-11-29T14%3A45%3A34Z#refactoring-a-component-to-typescript)

I've also added a README for the component.

## Testing Instructions
- Ensure all unit and e2e tests still pass
- In Storybook, confirm consumers like `ColorPalette`, `DuotonePicker`, and `GradientPicker` still work as expected.

chad1008 committed:
Rename to `.tsx` and remove `@ts-nocheck`

chad1008 committed:
Add `types.ts` file

chad1008 committed:
Add `CircularOptionPickerProps`

chad1008 committed:
Add `ButtonAction` typing

chad1008 committed:
Add `DropDownLinkAction` typing

chad1008 committed:
Add `Option` typing

chad1008 committed:
Add README

chad1008 committed:
Fix main component export

chad1008 committed:
update CHANGELOG

chad1008 commented on packages/components/src/circular-option-picker/types.ts:
```diff
@@ -0,0 +1,70 @@
+	'isPressed' | 'className'
+> & {
+	className?: string;
+	// TODO: Discuss pros/cons of something like `React.ComponentProps< typeof Tooltip >[ 'text' ];`
+	tooltipText?: string;
```
Something @ciampo and I discussed and wanted input from others on: in a case like this we can type `tooltipText` explicitly as a string, or tie it directly to the related prop type on `Tooltip` itself.

With the latter approach, if `Tooltip`'s `text` prop is ever retyped, `CircularOptionPicker` wouldn't need to be updated, and would stay in sink with the new change without any TS errors. On the other hand, if `Tooltip`'s `text` prop is ever retyped in an unexpected way, it might be helpful to see that in the form of a TS error to help avoid any unexpected side effects. 🤔

ciampo replied:
I thought about this, and I'd say that the safest thing to do here is to type this as `string`. As you mentioned, if `Tooltip` ever changed its APIs in a breaking way, we'd get a TS error in this file.

Let's keep `string` and delete the inline comment above

chad1008 replied:
✅  f1c051675e

ciampo commented on packages/components/src/circular-option-picker/types.ts:
```diff
@@ -0,0 +1,70 @@
+	isSelected: boolean;
+	// `icon` is explicitly defined as 'check' by CircleOptionPicker.Option
+	// and is not intended to be overridden.
+	// `size` relies on the `Icon` component's default size of `24` to fit
+	// `CircularOptionPicker`'s design, and should not be explicitly set.
```
Well spotted!

chad1008 replied:
Thanks!

ciampo commented on packages/components/src/circular-option-picker/README.md:
```diff
@@ -0,0 +1,81 @@
+# CircularOptionPicker
```
Thank you for adding a README!

Given the subcomponent structure, I think it would be beneficial to add props for the rest of the subcomponents, maybe something like 

```
# CircularOptionPicker

[...]

## Usage

[...]

## Props

[...]

## Subcomponents

### CircularOptionPicker.ButtonAction

description

#### Props

[...]

### CircularOptionPicker.DropdownLinkAction

description

#### Props

[...]

### CircularOptionPicker.Option

description

#### Props

[...]

```

What do you think?

chad1008 replied:
Sounds good. Subcomponents added to the README ✅, happy to make any additional revisions/improvements.

ciampo reviewed (commented):
Nicely done!

✅  No runtime changes
✅  Types look good to me
✅  Did a quick smoke test in `GradientPicker`'s Storybook example

Apart from the inline comments, would you be able to add a simple Storybook example? Nothing too complex, just a "standard" default example to show the auto-generated docs and test the component in isolation.

chad1008 committed:
remove TODO comment

chad1008 committed:
Add subcomponents to README, clean up whitespace

chad1008 committed:
Mark `DropdownLinkActionProps.buttonProps` as optional

chad1008 committed:
Add stories

chad1008 commented:
Thanks for the review @ciampo! I've updated the README with the subcomponents, and I've also added a story. While I was there I added stories to demonstrate the two subcomponents, in part because I wanted to experiment with `DropdownLinkAction` for my own edification. Let me know what you think 😄

chad1008 commented:
Shoot, I almost forgot: quick question about formatting the code examples in Storybook: I've used an array of options and then `map`ed them into suitable `Option` components... this works fine but leads to the `options` prop in the code examples not being formatted nicely.

Is there a trick to force some indentation there, or would we be better off just hardcoding those `Options` subcomponents so we can just explicitly format/indent as we see fit?

chad1008 commented on packages/components/src/circular-option-picker/stories/index.tsx:
```diff
@@ -0,0 +1,86 @@
+	{ color: '#0af', name: 'Blue' },
+];
+
+// TODO: ask if there's a more elegant way to access a state handler from a second story
+let resetColor: React.Dispatch< React.SetStateAction< number | undefined > >;
```
The template story doesn't include a 'clear' button, but the second story (`With ButtonAction`) does, and needs to modify a bit of state from the template. Is there a better way to make the state setter from the template accessible to other stories?

ciampo replied:
I fiddle a bit with this, and I think that the better way to achieve this is to use Storybook [decorators](https://storybook.js.org/docs/react/writing-stories/decorators) and React context, so that all components involved can share the same `color` state

I also made a couple of improvements:

- instead of using the color index to store the selected color, I chaged the logic to use the color value
- I made non-default stories inherit the default story's args 

<details>

<summary>Like this</summary>


```diff
diff --git a/packages/components/src/circular-option-picker/stories/index.tsx b/packages/components/src/circular-option-picker/stories/index.tsx
index eca64f905e..27eb9215dd 100644
--- a/packages/components/src/circular-option-picker/stories/index.tsx
+++ b/packages/components/src/circular-option-picker/stories/index.tsx
@@ -5,12 +5,17 @@ import type { ComponentMeta, ComponentStory } from '@storybook/react';
 /**
  * WordPress dependencies
  */
-import { useState } from '@wordpress/element';
+import { useState, createContext, useContext } from '@wordpress/element';
 /**
  * Internal dependencies
  */
 import CircularOptionPicker from '..';
 
+const CircularOptionPickerStoryContext = createContext< {
+	currentColor?: string;
+	setCurrentColor?: ( v: string | undefined ) => void;
+} >( {} );
+
 const meta: ComponentMeta< typeof CircularOptionPicker > = {
 	title: 'Components/CircularOptionPicker',
 	component: CircularOptionPicker,
@@ -25,6 +30,23 @@ const meta: ComponentMeta< typeof CircularOptionPicker > = {
 		controls: { expanded: true },
 		docs: { source: { state: 'open' } },
 	},
+	decorators: [
+		// Share current color state between main component, `actions` and `options`
+		( Story ) => {
+			const [ currentColor, setCurrentColor ] = useState< string >();
+
+			return (
+				<CircularOptionPickerStoryContext.Provider
+					value={ {
+						currentColor,
+						setCurrentColor,
+					} }
+				>
+					<Story />
+				</CircularOptionPickerStoryContext.Provider>
+			);
+		},
+	],
 };
 export default meta;
 
@@ -34,14 +56,12 @@ const colors = [
 	{ color: '#0af', name: 'Blue' },
 ];
 
-// TODO: ask if there's a more elegant way to access a state handler from a second story
-let resetColor: React.Dispatch< React.SetStateAction< number | undefined > >;
-
-const Template: ComponentStory< typeof CircularOptionPicker > = ( props ) => {
-	const [ currentColor, setCurrentColor ] = useState< number >();
-	resetColor = setCurrentColor;
+const DefaultOptions = () => {
+	const { currentColor, setCurrentColor } = useContext(
+		CircularOptionPickerStoryContext
+	);
 
-	const colorOptions = (
+	return (
 		<>
 			{ colors.map( ( { color, name }, index ) => {
 				return (
@@ -49,34 +69,47 @@ const Template: ComponentStory< typeof CircularOptionPicker > = ( props ) => {
 						key={ `${ color }-${ index }` }
 						tooltipText={ name }
 						style={ { backgroundColor: color, color } }
-						isSelected={ index === currentColor }
-						onClick={ () => setCurrentColor( index ) }
+						isSelected={ color === currentColor }
+						onClick={ () => {
+							setCurrentColor?.( color );
+						} }
 						aria-label={ name }
 					/>
 				);
 			} ) }
 		</>
 	);
-	return <CircularOptionPicker { ...props } options={ colorOptions } />;
 };
 
-export const Default = Template.bind( {} );
-Default.args = { actions: <></> };
+const DefaultActions = () => {
+	const { setCurrentColor } = useContext( CircularOptionPickerStoryContext );
 
-export const WithButtonAction = Template.bind( {} );
-WithButtonAction.args = {
-	actions: (
+	return (
 		<CircularOptionPicker.ButtonAction
-			onClick={ () => resetColor( undefined ) }
+			onClick={ () => setCurrentColor?.( undefined ) }
 		>
 			{ 'Clear' }
 		</CircularOptionPicker.ButtonAction>
-	),
+	);
+};
+
+const Template: ComponentStory< typeof CircularOptionPicker > = ( props ) => (
+	<CircularOptionPicker { ...props } />
+);
+
+export const Default = Template.bind( {} );
+Default.args = { actions: <></>, options: <DefaultOptions /> };
+
+export const WithButtonAction = Template.bind( {} );
+WithButtonAction.args = {
+	...Default.args,
+	actions: <DefaultActions />,
 };
 WithButtonAction.storyName = 'With ButtonAction';
 
 export const WithDropdownLinkAction = Template.bind( {} );
 WithDropdownLinkAction.args = {
+	...Default.args,
 	actions: (
 		<CircularOptionPicker.DropdownLinkAction
 			dropdownProps={ {
```

</details>

chad1008 replied:
This is fancy 😆 Applied: a1d1eb7cec

chad1008 replied:
Hm. Actually, applying this patch seems to have broken the examples on the story's Docs tab. 🤔 Looking into it now.

ciampo replied:
This should fix it 

```diff
diff --git a/packages/components/src/circular-option-picker/stories/index.tsx b/packages/components/src/circular-option-picker/stories/index.tsx
index b23f694b7d..db3943c9fa 100644
--- a/packages/components/src/circular-option-picker/stories/index.tsx
+++ b/packages/components/src/circular-option-picker/stories/index.tsx
@@ -27,7 +27,7 @@ const meta: ComponentMeta< typeof CircularOptionPicker > = {
 	parameters: {
 		actions: { argTypesRegex: '^on.*' },
 		controls: { expanded: true },
-		docs: { source: { state: 'open' } },
+		docs: { source: { state: 'open', excludeDecorators: true } },
 	},
 	decorators: [
 		// Share current color state between main component, `actions` and `options`
```

Source: https://github.com/storybookjs/storybook/issues/12022#issuecomment-859540813

chad1008 replied:
For historical context, the docs are currently displaying the following for me:

<img width="356" alt="image" src="https://user-images.githubusercontent.com/13856531/219148625-d80916b7-67ee-490b-9877-1df8f1ce00e1.png">

There appear to be two issues:

1. The decorator's code is rendering, but instead of showing the actual context provider, the component name is coming up as `[object Object]` and the `value` prop is being displayed directly.
2. Instead of rendering the actual story, a `<No Display Name>` component is being displayed.

Solutions:
1. I don't think the decorator likes the state management placed directly inside the `decorator` declaration. If I extract the state management out into a separate component before hand, it can actually render the Provider:
<img width="429" alt="image" src="https://user-images.githubusercontent.com/13856531/219151524-848f5d79-ccef-4e93-8812-00f9065394b5.png">

<details>

<summary>Example</summary>


```diff
diff --git a/packages/components/src/circular-option-picker/stories/index.tsx b/packages/components/src/circular-option-picker/stories/index.tsx
index b23f694b7d..cc1f1d344f 100644
--- a/packages/components/src/circular-option-picker/stories/index.tsx
+++ b/packages/components/src/circular-option-picker/stories/index.tsx
@@ -16,6 +16,25 @@ const CircularOptionPickerStoryContext = createContext< {
 	setCurrentColor?: ( v: string | undefined ) => void;
 } >( {} );
 
+const CircularOptionPickerStoryContextProvider = ( {
+	children,
+}: {
+	children: React.ReactNode;
+} ) => {
+	const [ currentColor, setCurrentColor ] = useState< string >();
+
+	return (
+		<CircularOptionPickerStoryContext.Provider
+			value={ {
+				currentColor,
+				setCurrentColor,
+			} }
+		>
+			{ children }
+		</CircularOptionPickerStoryContext.Provider>
+	);
+};
+
 const meta: ComponentMeta< typeof CircularOptionPicker > = {
 	title: 'Components/CircularOptionPicker',
 	component: CircularOptionPicker,
@@ -32,17 +51,10 @@ const meta: ComponentMeta< typeof CircularOptionPicker > = {
 	decorators: [
 		// Share current color state between main component, `actions` and `options`
 		( Story ) => {
-			const [ currentColor, setCurrentColor ] = useState< string >();
-
 			return (
-				<CircularOptionPickerStoryContext.Provider
-					value={ {
-						currentColor,
-						setCurrentColor,
-					} }
-				>
+				<CircularOptionPickerStoryContextProvider>
 					<Story />
-				</CircularOptionPickerStoryContext.Provider>
+				</CircularOptionPickerStoryContextProvider>
 			);
 		},
 	],

```

</details>

2. It looks like passing `<Story />` as the `children` prop of the decorator's return value [can result in that `<No Display Name>` issue](https://github.com/storybookjs/storybook/issues/12596#issuecomment-723440097). Passing `{Story()}` instead helps (this screenshot has also has the above changes as well):

<img width="407" alt="image" src="https://user-images.githubusercontent.com/13856531/219156640-eb7810cd-cf87-49ad-b609-4ae2fae38b64.png">

I haven't applied these in case this is something you've seen before and you have a different recommendation?

ciampo replied:
Have you already tried applying [this suggestion](https://github.com/WordPress/gutenberg/pull/47937#discussion_r1107713562)?

When I tried it on my machine, it seemed to fix the issue, only showing the actual story components in the snippet

chad1008 replied:
Ah - looks like we posted within a couple minutes of each other and I missed your reply. That does look a lot better!

chad1008 replied:
ha! nope. Looking at the timestamps it looks like you shared that juuuust before I dropped my wall of text there, so I missed it. 495a91128c

ciampo replied:
I thought so! Anyway, glad it worked :)

chad1008 commented on packages/components/src/circular-option-picker/stories/index.tsx:
```diff
@@ -0,0 +1,86 @@
+		controls: { expanded: true },
+		docs: { source: { state: 'open' } },
+	},
+};
```
I noticed the `actions` control in Storybook is currently displaying a gnarly object that isn't really useful... I wasn't 100% sure what caused that, but we can avoid it by setting that control type to `null`. What do you think?

ciampo replied:
Yup, let's set both `actions` and `options` controls to `null` — it would be impossible anyway to define a React component in the Storybook input field, plus we're also already passing those props in the examples.

We can also set the `children` prop's controls to be of type `string`, so that at least it's possible to play with the `children` prop in Storybook

chad1008 replied:
Done! e1d2aaeb2c

chad1008 replied:
✅ e1d2aaeb2c

ciampo commented on packages/components/src/circular-option-picker/README.md:
```diff
@@ -0,0 +1,147 @@
+# CircularOptionPicker
```
Let's format the component names to use monospace font

```suggestion
# `CircularOptionPicker`
```

Let's apply this throughout the README

chad1008 replied:
👍 eb514e0632

ciampo commented on packages/components/src/circular-option-picker/README.md:
```diff
@@ -0,0 +1,147 @@
+
+### actions
+
+The action(s) to be rendered after the options, such as a 'clear' button as seen in `ColorPalette'.
```
Mismatched closing tick

```suggestion
The action(s) to be rendered after the options, such as a 'clear' button as seen in `ColorPalette`.
```

chad1008 replied:
Thank you! eb514e0632

ciampo commented on packages/components/src/circular-option-picker/README.md:
```diff
@@ -0,0 +1,147 @@
+A CSS class to apply to the wrapper element.
+
+- Type: `string`
+- Required: No
```
Nit: let's also format prop names to use a monospace font — and while we're at it, we can also follow the newer format that we've been promoting for prop docs (ie. having the prop type next to the prop name)

```suggestion
### `className`: `string`

A CSS class to apply to the wrapper element.

- Required: No
```

Let's apply these changes throughout the README

chad1008 replied:
✅ eb514e0632

ciampo commented on packages/components/src/circular-option-picker/stories/index.tsx:
```diff
@@ -0,0 +1,86 @@
+
+const meta: ComponentMeta< typeof CircularOptionPicker > = {
+	title: 'Components/CircularOptionPicker',
+	component: CircularOptionPicker,
```
Let's also add the subcomponents via the `subcomponents` property.

In order to allow Storybook to extract types correctly from the subcomponents, we'll have to export them individually as well

<details>

<summary>Something like this</summary>

```diff
diff --git a/packages/components/src/circular-option-picker/index.tsx b/packages/components/src/circular-option-picker/index.tsx
index f03f969f69..6d98ed2a6a 100644
--- a/packages/components/src/circular-option-picker/index.tsx
+++ b/packages/components/src/circular-option-picker/index.tsx
@@ -22,7 +22,7 @@ import type {
 import type { WordPressComponentProps } from '../ui/context';
 import type { ButtonAsButtonProps } from '../button/types';
 
-function Option( {
+export function Option( {
 	className,
 	isSelected,
 	selectedIconProps,
@@ -58,7 +58,7 @@ function Option( {
 	);
 }
 
-function DropdownLinkAction( {
+export function DropdownLinkAction( {
 	buttonProps,
 	className,
 	dropdownProps,
@@ -86,7 +86,7 @@ function DropdownLinkAction( {
 	);
 }
 
-function ButtonAction( {
+export function ButtonAction( {
 	className,
 	children,
 	...additionalProps
diff --git a/packages/components/src/circular-option-picker/stories/index.tsx b/packages/components/src/circular-option-picker/stories/index.tsx
index 8e598e9533..87833298a4 100644
--- a/packages/components/src/circular-option-picker/stories/index.tsx
+++ b/packages/components/src/circular-option-picker/stories/index.tsx
@@ -9,7 +9,13 @@ import { useState, createContext, useContext } from '@wordpress/element';
 /**
  * Internal dependencies
  */
-import CircularOptionPicker from '..';
+import {
+	default as CircularOptionPicker,
+	Option,
+	ButtonAction,
+	DropdownLinkAction,
+} from '..';
+
 const CircularOptionPickerStoryContext = createContext< {
 	currentColor?: string;
 	setCurrentColor?: ( v: string | undefined ) => void;
@@ -18,6 +24,11 @@ const CircularOptionPickerStoryContext = createContext< {
 const meta: ComponentMeta< typeof CircularOptionPicker > = {
 	title: 'Components/CircularOptionPicker',
 	component: CircularOptionPicker,
+	subcomponents: {
+		'CircularOptionPicker.Option': Option,
+		'CircularOptionPicker.ButtonAction': ButtonAction,
+		'CircularOptionPicker.DropdownLinkAction': DropdownLinkAction,
+	},
 	parameters: {
 		actions: { argTypesRegex: '^on.*' },
 		controls: { expanded: true },
```

</details>

chad1008 replied:
✅ 25a4801c42

ciampo commented on packages/components/src/circular-option-picker/README.md:
```diff
@@ -0,0 +1,147 @@
+# CircularOptionPicker
+
+CircularOptionPicker is a component that displays a set of options as circular buttons.
```
Let's mention that the component is currently not exported

```suggestion
<div class="callout callout-alert">
This component is not exported, and therefore can only be used internally to the `@wordpress/components` package.
</div>

`CircularOptionPicker` is a component that displays a set of options as circular buttons.
```

chad1008 replied:
✅ 76d19c755e

ciampo commented on packages/components/src/circular-option-picker/README.md:
````diff
@@ -0,0 +1,147 @@
+## Usage
+
+```jsx
+import { CircularOptionPicker } from '@wordpress/components';
````
For the same reason as the previous comment: 

```suggestion
import { CircularOptionPicker } from '../circular-option-picker';
```

ciampo commented on packages/components/src/circular-option-picker/index.tsx:
```diff
@@ -93,7 +105,7 @@ function ButtonAction( props ) {
 }
 
-export default function CircularOptionPicker( props ) {
+function CircularOptionPicker( props: CircularOptionPickerProps ) {
```
We should add a JSDoc comment to `CircularOptionPicker` like we do for other components, in sync with the contents of the README file

chad1008 replied:
✅ c750820a5e

ciampo reviewed (commented):
Thank you for working on the README and the Storybook example.

I've left inline comments, but overall changes are looking great 🎉 

> ...this leads to the options prop in the code examples not being formatted nicely.
> Is there a trick to force some indentation there, or would we be better off just hardcoding those Options subcomponents so we can just explicitly format/indent as we see fit?

I'm not sure I understand what you mean  — how would I be able to see the issues with the formatting?

chad1008 committed:
Update Storybook `argTypes`

chad1008 committed:
Apply decorators and React context to stories

chad1008 committed:
README formatting

ciampo commented on packages/components/src/circular-option-picker/stories/index.tsx:
```diff
@@ -0,0 +1,124 @@
+);
+
+export const Default = Template.bind( {} );
+Default.args = { actions: <></>, options: <DefaultOptions /> };
```
Since the `actions` prop is optional, we can just avoid passing it instead of passing an empty fragment

```suggestion
Default.args = { options: <DefaultOptions /> };
```

chad1008 committed:
Add subcomponents to story

chad1008 committed:
Add `not exported` note

chad1008 committed:
Fix import example in README

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

chad1008 committed:
Remove unneeded `actions` prop from story

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

chad1008 committed:
JSDoc FTW

chad1008 commented:
> I'm not sure I understand what you mean — how would I be able to see the issues with the formatting?

Sorry, I wasn't clear in my description. On the Docs tab in Storybook, the code example was rendering with the `options` prop listed on a single line, rather than placing each `Option` subcomponent on it's own line, nicely indented, etc. I wasn't sure if what caused that or if there was a good fix to make the example more readable. With some of the more recent changes, this isn't happening any longer, but it may still be a thing once we sort out the current issues with the decorated we added.

I think I'm caught up on existing feedback. I'll address the CHANGELOG conflict as a final step when the other ducks are all in a row. Thank you again for all of the detailed feedback!

chad1008 committed:
Exclude decorator from Storybook docs

ciampo commented on packages/components/src/circular-option-picker/index.tsx:
```diff
@@ -118,3 +130,51 @@ export default function CircularOptionPicker( props ) {
 CircularOptionPicker.ButtonAction = ButtonAction;
 CircularOptionPicker.DropdownLinkAction = DropdownLinkAction;
+
+/**
```
In order to be associated to the `CircularOptionPicker` component, this JSDoc comment needs to be moved next to where the component is declared (basically on top of `function CircularOptionPicke....`)

chad1008 replied:
🤦 Fixed, thank you.

ciampo reviewed (approved):
Great work here and thank you for the patience 🚀

chad1008 committed:
move JSdoc example to proper position

chad1008 committed:
typo

chad1008 commented:
Thank you for all of the thoughtful help and feedback @ciampo!
